paddle/fluid/operators/matmul_v2_op.cc

```diff
@@ -62,10 +62,15 @@ class MatMulV2Op : public framework::OperatorWithKernel {
     }
 
     std::vector<int64_t> new_dims;
-    if (ndims_x >= ndims_y) {
+    if (ndims_x > ndims_y) {
       new_dims.assign(dims_x.begin(), dims_x.end() - 2);
-    } else {
+    } else if (ndims_x < ndims_y) {
       new_dims.assign(dims_y.begin(), dims_y.end() - 2);
+    } else {
+      new_dims.reserve(ndims_x);
+      for (size_t i = 0; i < ndims_x - 2; ++i) {
+        new_dims.push_back(std::max(dims_x[i], dims_y[i]));
+      }
```

Comment:
Are you sure that this is the correct behavior? This significantly change the ops behavior by adding broadcasting. Moreover all already written software already relies on fact that the `dims_x` is taken when X input tensor has bigger rank.  
Besides above note you should probably check whether the processed axes are broadcastable according to NumPy rules.

Reply:
I am sure about that.
PaddlePaddle matmul_v2 native kernel does that in Compute() method instead of in InferShape()(I have no idea why) and this incorrect behavior of InferShape resulted in BF16 tests failure.
![image](https://user-images.githubusercontent.com/62569058/126171951-8f4670c0-6ceb-4eb1-8388-5a3e2f346936.png)
They also have test for broadcasting both X and Y:
![image](https://user-images.githubusercontent.com/62569058/126171668-a6e846bd-bd51-4e43-bdbf-3508cd0b741e.png)